riptable/rt_str.py

```diff
@@ -915,5 +876,87 @@ def char(self, position: Union[int, List[int], np.ndarray]):
     nb_char_par = _njit_par(_nb_char)
 
 
+def _populate_wrappers(cls):
+    """
+    Decorator for the CatString class which populates the string methods and sets the defaults for
+    filtered value fills.
+    """
+    functions = dict(inspect.getmembers(FAString, inspect.isfunction))
+    properties = dict(inspect.getmembers(FAString, lambda o: isinstance(o, (cached_property, property))))
+    for name, filter_value_fill in [
+        ('upper', None),
+        ('lower', None),
+        ('reverse', None),
+        ('removetrailing', None),
+        ('strlen', INVALID_POINTER_32),
+        ('index_any_of', INVALID_POINTER_32),
+        ('index', INVALID_POINTER_32),
+        ('contains', False),
+        ('startswith', False),
+        ('endswith', False),
+        ('regex_match', False),
+        ('substr', None),
+        ('char', None)
+    ]:
+        if name in functions:
+            wrapper = cls._build_method(functions[name], filter_value_fill)
+        elif name in properties:
+            wrapper = cls._build_property(name, filter_value_fill)
+        else:
+            raise RuntimeError(f'{name} is not defined on FAString as a function or property')
+        setattr(cls, name, wrapper)
+    return cls
+
+
+@_populate_wrappers
+class CatString:
+    """
+    Provides access to FAString methods for Categoricals.
+    All string methods are wrappers of the FAString equivalent with
+    categorical re-expansion and option for how to fill filtered elements.
+    """
+
+    def __init__(self, cat):
+        from .rt_categorical import CategoryMode
+        self.cat = cat
+        if cat.category_mode == CategoryMode.StringArray:
+            string_list = cat.category_array
+            self.fastring = FAString(string_list)
+        else:
+            raise ValueError(f"Could not use .str in {CategoryMode(cat.category_mode).name}.")
+
+    def _convert_fastring_output(self, out, filter_fill_value):
+        from .rt_categorical import Categorical
+        if out.dtype.kind in ['S', 'U']:
+            out = Categorical(self.cat._fa.copy(), out, base_index=self.cat.base_index)

```

Comment:
I think the creation of the output ``Categorical`` or array (in the ``else`` branch) could probably be optimized a bit more to eliminate some work + temporary allocations. However, given the performance and correctness gains we'll get from the changes as-is, I think it's fine to leave those additional optimizations until some later time.

Reply:
Yes, I think if we had a `Categorical._from_nonunique_labels` method that didn't require use to build this interim one that would cut at least one copy.